### ISSUE TITLE
Switch console tools to cxxopts

### ISSUE THREAD

**Issue by lemire** [simdjson/simdjson#814]
We are using getopt which is old school, we should upgrade to something like this: https://github.com/jarro2783/cxxopts

**Comment by lemire** [comment]
I am calling for help on this issue. Nice little project. Remove getopt from simdjson and put something modern (not boost, please, not boost).

**Comment by jkeiser** [comment]
@lemire see #799 :)

**Comment by jkeiser** [comment]
But yes, we should use something that is not getopt, so we'll still need this!

**Comment by pps83** [comment]
Just add this?
https://github.com/powturbo/TurboPFor-Integer-Compression/blob/master/vs/getopt.h
https://github.com/powturbo/TurboPFor-Integer-Compression/blob/master/vs/getopt.c

**Comment by lemire** [comment]
@pps83 

getopt is ancient. Did you look at https://github.com/jarro2783/cxxopts
It looks much nicer. More C++-ish.

Instead of just patching, we should modernize the code.

**Comment by pps83** [comment]
> @pps83
> 
> getopt is ancient. Did you look at https://github.com/jarro2783/cxxopts
> It looks much nicer. More C++-ish.
> 
> Instead of just patching, we should modernize the code.

but if getopt already works, why would you spend time something that works?

**Comment by lemire** [comment]
> but if getopt already works, why would you spend time on something that works?

Hmmm... That's a reasonable question.

I think that the bulk of the work involved in fixing this problem involves importing a new file, making sure that everything work. The work of transforming our existing code, in comparison is small.

It is like when you change your dishwasher... you might as well clean the plumbing...

**Comment by jkeiser** [comment]
To be clear, the linked pr already borrows a public getopt.h.

**Comment by lemire** [comment]
I did notice. My larger point was that while at it we could have modernized the tools, but it is fine if we don't.

**Comment by jkeiser** [comment]
Cool. I would like modern command line too, I just wasn't in the mood to detour on it at the time.

**Comment by jkeiser** [comment]
cxxopts looks pretty good, too :)

**Comment by lshamis** [comment]
Consider https://github.com/CLIUtils/CLI11

**Comment by lemire** [comment]
@lshamis  How does it compare with  cxxopts?

cc @yoannlr

**Comment by h5rdly** [comment]
`cxxopts.hpp` is 2215 lines, `CLI11.hpp` is 8220 lines

**Comment by data-man** [comment]
@lemire 
CLI11 is more feature rich and has many built-in [validators](https://github.com/CLIUtils/CLI11#validators).

**Comment by lemire** [comment]
@data-man Point taken. I think that if you want to build rich command-line tools, CLI11 looks better. It has subcommands and the like.

**Comment by data-man** [comment]
I also propose use [nanobench](https://github.com/martinus/nanobench). :)

**Comment by lemire** [comment]
Closed via https://github.com/simdjson/simdjson/pull/904 

I will create a new PR for related work.

**Comment by jkeiser** [comment]
@lemire @data-man I was just trying to port parse to cxxopts (I abandoned it at the end of my timebox) and MAN, that's a lackluster command line argument tool. I've used better ones in every other language the past few years (Rust, Perl, Python, Ruby, C#). The lack of validation is a real problem IMO, you want to be able to say "-a" and give "haswell," "westmere," "fallback" as options, or "-s" and give "stage1" and "all" as the only options.
